[SPARK-25951][SQL] Ignore aliases for distributions and orderings

mgaido91 opened the pull request:
## What changes were proposed in this pull request?

When we canonicalize an `Expression`, we do not remove `Alias`. So two expressions which are the same but are renamed are considered to be semantically different. As we rely on semantic equality in order to check if the result is the same for 2 expressions, some optimizations - as showed in the JIRA - may fail to apply, eg. removing redundant shuffles, when a column is renamed.

The PR proposes to ignore `Alias`es when checking whether distributions and orderings are satisfied by introducing a new method `sameResults` which ignore `Alias`es.

Credit should be given to @maropu for the approach suggestion which follows #17400.

Closes #17400.

## How was this patch tested?

added UT

mgaido91 committed:
[SPARK-25951][SQL] Remove Alias when canonicalize

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4798/
Test PASSed.

SparkQA commented:
**[Test build #98521 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98521/testReport)** for PR 22957 at commit [`b566818`](https://github.com/apache/spark/commit/b566818bf17c2b69d06fed2e1500e608a25bb3b7).

SparkQA commented:
**[Test build #98521 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98521/testReport)** for PR 22957 at commit [`b566818`](https://github.com/apache/spark/commit/b566818bf17c2b69d06fed2e1500e608a25bb3b7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98521/
Test FAILed.

mgaido91 committed:
introduce sameResult

SparkQA commented:
**[Test build #98552 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98552/testReport)** for PR 22957 at commit [`2b00f35`](https://github.com/apache/spark/commit/2b00f351fbb8f33cf01b30297f3785956b1b3697).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4817/
Test PASSed.

SparkQA commented:
**[Test build #98552 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98552/testReport)** for PR 22957 at commit [`2b00f35`](https://github.com/apache/spark/commit/2b00f351fbb8f33cf01b30297f3785956b1b3697).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/98552/
Test PASSed.

rxin commented:
i didn't look at your new code, but is your old code safe? e.g. a project that depends on the new alias.

mgaido91 commented:
cc @cloud-fan @gatorsmile

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -195,14 +195,26 @@ abstract class Expression extends TreeNode[Expression] {
+   * different output expressions can evaluate to the same result as well (eg. when an expression
+   * is aliased).
+   */
+  def sameResult(other: Expression): Boolean = other match {
```
I know it's always safer to introduce a new API, does is it really necessary? In `Canonicalize`, we erase the name for attributes, I think it's reasonable to erase the name of `Alias`, as it doesn't affect the output.

cloud-fan replied:
"erase the name" can also mean remove `Alias`. If we can't clearly tell the difference between `semanticEquals` and `sameResult`, and give a guideline about using which one in which case, I think we should just update `semanticEquals`(i.e. `Canonicalize`).

mgaido91 replied:
remove `Alias` is not possible for the reason explained in https://github.com/apache/spark/pull/22957#issuecomment-436992955. In general, `semanticEquals` should be used when we want to replace an expression with another, while `sameResult` should be used in order to check that 2 expressions return the same output.

cloud-fan replied:
can you put it in the method doc(both `semanticEquals` and `sameResult`)? This makes sense to me.

mgaido91 replied:
Sure, thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -195,14 +195,26 @@ abstract class Expression extends TreeNode[Expression] {
+   */
+  def sameResult(other: Expression): Boolean = other match {
+    case a: Alias => sameResult(a.child)
+    case _ => this.semanticEquals(other)
```
can we also strip the alias of this here? so that we can mark `sameResult` as final.

mgaido91 replied:
I think it is doable, but I didn't want to put too many `match` where it was not needed. But if you prefer that way I can try and do that.

mgaido91 replied:
well, it needs to be overridden by `HashPartitioning` too, so since I am not able to make it final anyway, I don't think it is a good idea. Well, I can add a match on `HashPartitioning`too, but it doesn't seem a clean solution to me.

cloud-fan replied:
we can do
```
CleanupAliases.trimAliases(this) semanticEquals CleanupAliases.trimAliases(other)
```

mgaido91 committed:
address comment: add comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5451/
Test PASSed.

SparkQA commented:
**[Test build #99375 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99375/testReport)** for PR 22957 at commit [`0491249`](https://github.com/apache/spark/commit/049124971eb9d0f84eb98cc33c4cccd65d7a42b7).

mgaido91 committed:
address comment

SparkQA commented:
**[Test build #99376 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99376/testReport)** for PR 22957 at commit [`3831be0`](https://github.com/apache/spark/commit/3831be0aba7eaf83bb03152de129c11286290586).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5452/
Test PASSed.

mgaido91 committed:
improve comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5453/
Test PASSed.

SparkQA commented:
**[Test build #99377 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99377/testReport)** for PR 22957 at commit [`6c93e70`](https://github.com/apache/spark/commit/6c93e708df203cc5a2f3085340ed61bf5c8d90fe).

SparkQA commented:
**[Test build #99376 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99376/testReport)** for PR 22957 at commit [`3831be0`](https://github.com/apache/spark/commit/3831be0aba7eaf83bb03152de129c11286290586).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99376/
Test FAILed.

SparkQA commented:
**[Test build #99375 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99375/testReport)** for PR 22957 at commit [`0491249`](https://github.com/apache/spark/commit/049124971eb9d0f84eb98cc33c4cccd65d7a42b7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99375/
Test PASSed.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #99424 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99424/testReport)** for PR 22957 at commit [`6c93e70`](https://github.com/apache/spark/commit/6c93e708df203cc5a2f3085340ed61bf5c8d90fe).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5496/
Test PASSed.

SparkQA commented:
**[Test build #99424 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99424/testReport)** for PR 22957 at commit [`6c93e70`](https://github.com/apache/spark/commit/6c93e708df203cc5a2f3085340ed61bf5c8d90fe).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99424/
Test FAILed.

mgaido91 committed:
fix recursive aliases

SparkQA commented:
**[Test build #99446 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99446/testReport)** for PR 22957 at commit [`a306465`](https://github.com/apache/spark/commit/a306465d3fc0ef2c8ce43cfd14277d3aab4550e2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5517/
Test PASSed.

SparkQA commented:
**[Test build #99446 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99446/testReport)** for PR 22957 at commit [`a306465`](https://github.com/apache/spark/commit/a306465d3fc0ef2c8ce43cfd14277d3aab4550e2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99446/
Test FAILed.

mgaido91 commented:
retest this please

SparkQA commented:
**[Test build #99448 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99448/testReport)** for PR 22957 at commit [`a306465`](https://github.com/apache/spark/commit/a306465d3fc0ef2c8ce43cfd14277d3aab4550e2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5520/
Test PASSed.

SparkQA commented:
**[Test build #99449 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99449/testReport)** for PR 22957 at commit [`a306465`](https://github.com/apache/spark/commit/a306465d3fc0ef2c8ce43cfd14277d3aab4550e2).

SparkQA commented:
**[Test build #99448 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99448/testReport)** for PR 22957 at commit [`a306465`](https://github.com/apache/spark/commit/a306465d3fc0ef2c8ce43cfd14277d3aab4550e2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99448/
Test FAILed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -2542,10 +2542,10 @@ object EliminateUnions extends Rule[LogicalPlan] {
     e.transformDown {
-      case Alias(child, _) => child
-      case MultiAlias(child, _) => child
+      case Alias(child, _) => trimAliases(child)
```
what's going on here?

mgaido91 replied:
the point is that now this method removes only the first `Alias` it finds (and it doesn't go on recursively), which is the reason of the UT failure. Also checking the comment on the method it seems not the expected behavior of this method.

mgaido91 replied:
ah, I did a stupid thing here. So the problem is that: since it returns `child` for `this`, in transformDown we apply the rule to `child` children, instead of applying to `child` itself. So the problem here is with 2 consecutive `Alias`. Let me find a better fix.

mgaido91 replied:
just using `transformUp`solves the issue

mgaido91 committed:
fix trimAliases

SparkQA commented:
**[Test build #99462 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99462/testReport)** for PR 22957 at commit [`13aef71`](https://github.com/apache/spark/commit/13aef71df5662f4f8389ece87c4270597de93db5).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5530/
Test PASSed.

SparkQA commented:
**[Test build #99449 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99449/testReport)** for PR 22957 at commit [`a306465`](https://github.com/apache/spark/commit/a306465d3fc0ef2c8ce43cfd14277d3aab4550e2).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99449/
Test FAILed.

SparkQA commented:
**[Test build #99462 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99462/testReport)** for PR 22957 at commit [`13aef71`](https://github.com/apache/spark/commit/13aef71df5662f4f8389ece87c4270597de93db5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99462/
Test PASSed.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -780,6 +780,23 @@ class PlannerSuite extends SharedSQLContext {
     }
   }
+
+  test("SPARK-25951: avoid redundant shuffle on rename") {
```
can we have an end-to-end test as well?

viirya replied:
+1 if possible.

mgaido91 replied:
ah, good point and indeed very useful. In my previous tests I always used a very simple query to verify this and never the one reported in the JIRA. Now I tried that one and I realized that this fix is not very useful as of now, because in renaming like that in the `HashPatitioning` there is the `AttributeReference` to the `Alias`, rather than the `Alias` itself. Since that is the common case, the PR as it is now it is not very useful. If I won't be able to figure out a good way for that, I am going to close this. Thanks and sorry for the trouble.

mgaido91 replied:
@cloud-fan @viirya I added the test, but as I mentioned I had to do another change in order to make it working. Sorry for the mistake. I'd really appreciate if you could review it again. Thanks.

cloud-fan commented:
LGTM, cc @viirya as well

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -195,14 +195,35 @@ abstract class Expression extends TreeNode[Expression] {
 
+  /**
+   * Returns true when two expressions will always compute the same result, even if the output may
+   * be different, because of different names or similar differences.
```
I think here `output` is a bit confusing. Do we mean the output names?

viirya replied:
So sameResult returns if the evaluated results between two expressions are exactly the same?

mgaido91 replied:
yes, I mean: `sameResult` returns true if 2 expressions return the same data even though from plan perspective they are not the same (eg. the output name/exprIds is different as in this case), while `semanticEquals` ensure they are the same from plan perspective too. If you have better suggestions how to rephrase this, I am happy to improve it. Thanks.

viirya replied:
How about replace `output` with `output from plan perspective`?

```
Returns true when two expressions will always compute the same result, even if the output
from plan perspective may be different, because of different names or similar differences.
```

viirya commented:
This looks good to me. Just a comment about wording.

mgaido91 committed:
address comments

SparkQA commented:
**[Test build #99517 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99517/testReport)** for PR 22957 at commit [`5c6b9fc`](https://github.com/apache/spark/commit/5c6b9fc52d825c5c6c680d38f8ef8c621dded964).

mgaido91 committed:
fix import

SparkQA commented:
**[Test build #99517 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99517/testReport)** for PR 22957 at commit [`5c6b9fc`](https://github.com/apache/spark/commit/5c6b9fc52d825c5c6c680d38f8ef8c621dded964).
 * This patch **fails Scala style tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] with PredicateHelper `

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99517/
Test FAILed.

mgaido91 committed:
Merge branch 'master' of github.com:apache/spark into SPARK-25951

SparkQA commented:
**[Test build #99518 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99518/testReport)** for PR 22957 at commit [`6eee1e4`](https://github.com/apache/spark/commit/6eee1e4235633284a0824b500d6b95cedbce7b0b).

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5581/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
**[Test build #99518 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99518/testReport)** for PR 22957 at commit [`6eee1e4`](https://github.com/apache/spark/commit/6eee1e4235633284a0824b500d6b95cedbce7b0b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99518/
Test FAILed.

mgaido91 committed:
fix

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5586/
Test PASSed.

SparkQA commented:
**[Test build #99522 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99522/testReport)** for PR 22957 at commit [`5bca5e3`](https://github.com/apache/spark/commit/5bca5e348f9ee6d1f00a5f68666e991160498ed7).

SparkQA commented:
**[Test build #99522 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99522/testReport)** for PR 22957 at commit [`5bca5e3`](https://github.com/apache/spark/commit/5bca5e348f9ee6d1f00a5f68666e991160498ed7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99522/
Test PASSed.

gatorsmile commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -780,6 +780,23 @@ class PlannerSuite extends SharedSQLContext {
     }
   }
+
+  test("SPARK-25951: avoid redundant shuffle on rename") {
```
I think the test case coverage is not enough for such massive changes.

mgaido91 replied:
thanks @gatorsmile, I added also a negative case, but I don't think it is enough. Do you have some hints on cases to test? Thanks.

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala:
```diff
@@ -223,14 +223,35 @@ abstract class Expression extends TreeNode[Expression] {
+   * same). It should not be used (and `semanticEquals` should be used instead) when comparing if 2
+   * expressions are the same and one can replace the other.
+   */
+  final def sameResult(other: Expression): Boolean =
```
Add unit test cases to SubexpressionEliminationSuite?

mgaido91 replied:
thanks, added

mgaido91 committed:
add tests

SparkQA commented:
**[Test build #99618 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99618/testReport)** for PR 22957 at commit [`1f797df`](https://github.com/apache/spark/commit/1f797dff9c28adccec3f32acaf361e3b94596ba0).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5672/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:
```diff
@@ -145,9 +145,14 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
     assert(requiredChildDistributions.length == children.length)
     assert(requiredChildOrderings.length == children.length)
 
+    val aliasMap = AttributeMap[Expression](children.flatMap(_.expressions.collect {
```
is it safe to do so? I think we should only collect aliases from operators with a project list, i.e. `Project`, `Aggregate`, `Window`.

mgaido91 replied:
I think it is. We are only checking the presence of aliases. In particular, we are collecting all the aliases which are defined in the previous operator. The solution you are suggesting works too IMHO and restricts the scope, but I am not sure it is a good thing, because I see no harm in doing it for other operators: simply they won't contain aliases; while I do see some issues in the maintenance of the "whitelist" of operators you are suggesting (we may miss some now or forget to update later...)

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:
```diff
@@ -145,9 +145,14 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
     children = children.zip(requiredChildDistributions).map {
-      case (child, distribution) if child.outputPartitioning.satisfies(distribution) =>
+      case (child, distribution) if child.outputPartitioning.satisfies(
+          distribution.mapExpressions(replaceAlias(_, aliasMap))) =>
```
instead of doing it here, shall we deal with alias in `SparkPlan.outputPartitioning` for operators with a project list?

mgaido91 replied:
this is not dealing with the aliases in the `outputPartitioning` but with the ones in the `requiredChildDistribution`. Anyway, I wouldn't do it there, because this would mean moving also the logic for collecting the aliases from the children there, which seems to me an operations which belong to a rule/transforming operator, rather than to the plan operator itself (eg. now these methods are in `PredicateHelper`...).

cloud-fan replied:
As an example, `ProjectExec.outputPartitioning` can be wrong, as it doesn't consider the aliases in the project list. I think it's clearer to adjust the `outputPartitioning` there, instead of dealing with it in a rule. What if we have more rules need to check `outputPartitioning` and `requiredChildDistribution`?

cloud-fan replied:
for example, `relation[a, b]`'s output partitioning is `[hash partition a, hash partition b]`, and `Project(a as c, b, relation)`'s output partitioning should be `[hash partition c, hash partition b]`.

What do you mean by `But ProjectExec.outputPartitioning cannot contain a reference to the aliases`?

mgaido91 replied:
ah I see now what you mean. I am not sure what you are suggesting is feasible. Imagine that in your example the Project is: `Project(a as c, a as d, b, relation)`. What should the output partitioning be?

> What do you mean by ...

I meant that when we collect the alias for `a as c`, we are mapping all the attr references of `c` with `a as c` here. In the `outputPartitioning`, there will never be an occurrence of a reference to `c`, but only references to `a`.

cloud-fan replied:
For `Project(a as c, a as d, b, relation)`, I think the `outputPartitioning` should be `[hash part c, hash part d, hash part b]`. The point is, we should not report an output partitioning whose attribute is not even in the current plan's output.

mgaido91 replied:
I don't think that is right: that would cause the shuffle to happen for every plan which is hashed by both `[hash part c, hash part b]` and `[hash part d, hash part b]` (and also `[hash part a, hash part b]`). I think that if we want to go that way, we'd need a set of equivalent outputPatitioning

cloud-fan replied:
seems we are not on the same page...

Let's make the example clearer. Assuming a `relation[a ,b]`'s partitioning is `hash(a, b)`, then `Project(a as c, a as d, b, relation)`'s partitioning should be `[hash(c, b), hash(d, b)]`. It's like a flatMap.

mgaido91 replied:
Ok, I think I got it now, sorry, I didn't understand :) yes I think this is doable then. Let me try and do that, thanks.

SparkQA commented:
**[Test build #99618 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99618/testReport)** for PR 22957 at commit [`1f797df`](https://github.com/apache/spark/commit/1f797dff9c28adccec3f32acaf361e3b94596ba0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

mgaido91 committed:
switch approach: update outputPartitioning

SparkQA commented:
**[Test build #99682 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99682/testReport)** for PR 22957 at commit [`bf1d04a`](https://github.com/apache/spark/commit/bf1d04a819855737d1096b61b1c3d46010f50dee).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5735/
Test PASSed.

mgaido91 committed:
fix ut error

SparkQA commented:
**[Test build #99713 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99713/testReport)** for PR 22957 at commit [`e4f617f`](https://github.com/apache/spark/commit/e4f617fc7e47d7c49f3d773ac2d91c5508c0a239).

SparkQA commented:
**[Test build #99713 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99713/testReport)** for PR 22957 at commit [`e4f617f`](https://github.com/apache/spark/commit/e4f617fc7e47d7c49f3d773ac2d91c5508c0a239).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99713/
Test PASSed.

mgaido91 committed:
Add test suite dedicated to partitioning

mgaido91 commented:
@cloud-fan @gatorsmile I updated the PR according to the previous suggestions and added a new dedicated test suite.
May you please review this again? Thanks.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5858/
Test PASSed.

SparkQA commented:
**[Test build #99834 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99834/testReport)** for PR 22957 at commit [`bca3e87`](https://github.com/apache/spark/commit/bca3e873b3193bcccd68b4e042fec5c985718d5b).

SparkQA commented:
**[Test build #99834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99834/testReport)** for PR 22957 at commit [`bca3e87`](https://github.com/apache/spark/commit/bca3e873b3193bcccd68b4e042fec5c985718d5b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99834/
Test FAILed.

mgaido91 committed:
fix UT failures

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/5864/
Test PASSed.

SparkQA commented:
**[Test build #99840 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99840/testReport)** for PR 22957 at commit [`0f68a41`](https://github.com/apache/spark/commit/0f68a41beb15da35c6839ecef18ec69d29e133dc).

SparkQA commented:
**[Test build #99840 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/99840/testReport)** for PR 22957 at commit [`0f68a41`](https://github.com/apache/spark/commit/0f68a41beb15da35c6839ecef18ec69d29e133dc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/99840/
Test FAILed.

mgaido91 commented:
retest this please

mgaido91 commented:
@cloud-fan @gatorsmile any more comments on this? Thanks.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:
```diff
@@ -1549,6 +1551,38 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       Seq(Row("Amsterdam")))
   }
 
+  test("SPARK-25951: avoid redundant shuffle on rename") {
```
I'm a little worried about introducing such a big change for such a corner case.

Looking at this test case, it seems updating the `outputPartitioning` according to the project list is good enough?

cc @maryannxue

mgaido91 replied:
> Looking at this test case, it seems updating the outputPartitioning according to the project list is good enough?

Basically, that is what is done. The diff is quite big, that's true, but mostly are tests. Is there something specific you are worried about?

cloud-fan replied:
I'm worried about `Expression.sameResult`. Is it still useful after updating the outputPartitioning according to the project list?

mgaido91 replied:
it is still useful because it may happen a used writes something like:
```
select ... group by a as b, ...
```
(please notice that despite it makes no sense written in SQL, maybe with other APIs is easier to happen).

And especially it is useful when updating the `outputPartitioning` in order to check whether an expression produces the same result of another.

mgaido91 commented:
anymore comments on this?

mgaido91 committed:
Merge branch 'master' into SPARK-25951

mgaido91 committed:
fix merge

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7583/
Test PASSed.

SparkQA commented:
**[Test build #101894 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101894/testReport)** for PR 22957 at commit [`9af290e`](https://github.com/apache/spark/commit/9af290ef546852f895672d65046fbe779ed09efb).

SparkQA commented:
**[Test build #101894 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101894/testReport)** for PR 22957 at commit [`9af290e`](https://github.com/apache/spark/commit/9af290ef546852f895672d65046fbe779ed09efb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101894/
Test PASSed.

cloud-fan commented:
I don't really agree with this approach. I think we should just propagate partitioning and ordering through Project/Aggregat/Window and any other operators that have project list.

mgaido91 commented:
@cloud-fan thanks for your comment. You mean we should run `updatePartitioningWithNewOutput` only on those operators or do you have something else in mind? Thanks.

cloud-fan commented:
I'm not sure if all the operators have same logic of propagating child partitioning and ordering. In general we want every operator to propagate children's partitioning and ordering if possible.

mgaido91 commented:
@cloud-fan sorry, I am not sure I am 100% getting what you mean, may I please ask you to elaborate a bit more? Thanks.

cloud-fan commented:
by default `SparkPlan.outputPartitioning` is `UnknownPartitioning`, this is safe but is sub-optimal for some operators like `ProjectExec`.

`ProjectExec.outputPartitioning` just returns `child.outputPartitioning`, which is better but still sub-optimal. If we have `a as c`, the partitioning info of column `c` is not propagated.

cloud-fan commented:
And I think this is the root cause of the problem, we don't need `Expression.sameResult `

mgaido91 commented:
yes, that is what this PR does, it propagates the partitioning info for column `c`. So I don't understand what your concerns are about the current approach. In particular that is performed only for Aggregates, Project and Window (we can extend this list in the future if we see more operators can do the same).

mgaido91 commented:
ah, I see now your point. So let me elaborate it a bit more, probably I was not clear. There are 2 cases which are addressed here:
 1. `select * from t1 join (select a as b, count(*) from t2 group by a) on t1.b=t2.b`: this is the case we are referring to and the `sameResult` usage in the check for the Partitionings/Orderings is not needed;
 2. there is a weirder case, when the alias is used in the ORDER BY/GROUP BY clause (eg. `select ... group by a as b, ...`). For this case we need the `sameResult` there.

So there are 2 cases fixed in 2 different ways. If you prefer, we can focus only on the first case for this PR and remove the usage of `Expression.sameResult` in checking the Partionings/Orderings. What do you think?

mgaido91 commented:
let me try and do that, thanks

mgaido91 committed:
Merge branch 'master' of github.com:apache/spark into SPARK-25951

mgaido91 committed:
address comments

mgaido91 committed:
cleanup

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7645/
Test PASSed.

SparkQA commented:
**[Test build #101964 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101964/testReport)** for PR 22957 at commit [`205f1b7`](https://github.com/apache/spark/commit/205f1b7f6a01ecab3d8ced4ea98c1bd17bd2960f).

SparkQA commented:
**[Test build #101964 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101964/testReport)** for PR 22957 at commit [`205f1b7`](https://github.com/apache/spark/commit/205f1b7f6a01ecab3d8ced4ea98c1bd17bd2960f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101964/
Test FAILed.

mgaido91 commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7649/
Test PASSed.

SparkQA commented:
**[Test build #101968 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101968/testReport)** for PR 22957 at commit [`205f1b7`](https://github.com/apache/spark/commit/205f1b7f6a01ecab3d8ced4ea98c1bd17bd2960f).

SparkQA commented:
**[Test build #101968 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101968/testReport)** for PR 22957 at commit [`205f1b7`](https://github.com/apache/spark/commit/205f1b7f6a01ecab3d8ced4ea98c1bd17bd2960f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/101968/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -186,6 +187,69 @@ trait Partitioning {
+              partitionings.map {
+                case hp: HashPartitioning if hp.references.contains(toReplace) =>
+                  UnknownPartitioning(hp.numPartitions)
+                case rp: RangePartitioning if rp.references.contains(toReplace) =>
```
since we need to handle different partitioning differently, shall we add a method to `Partitioning`?

mgaido91 committed:
address comment

SparkQA commented:
**[Test build #101982 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/101982/testReport)** for PR 22957 at commit [`f47d5df`](https://github.com/apache/spark/commit/f47d5df0336aa0a61356502140968979e2ff6d3c).

HeartSaVioR commented:
Sorry about jumping in, but looks like we missed #17400 (SPARK-19981) which looks like containing the fix what @cloud-fan suggests - reviewers just lost focus on that. For me #17400 looks very concise, so curious how this patch is different from #17400.

mgaido91 commented:
@HeartSaVioR thanks for pointing that out! Yes, that fix is similar to this, the main difference between these 2 PRs is that this one handles also the case when we have `select a as b, a, ...` and then we have the partitioning on `a`, while #17400 doesn't (and hence may introduce a regression). Let me cc @maropu too here, since he worked on this and may have comments/suggestion or he can update his PR to handle that case and we can focus on that PR.

maropu commented:
@mgaido91 Thank for letting me know! Could you fix this pr based on my fix? If you can, its ok to close my pr in #17400.

mgaido91 commented:
@maropu thanks for checking this. Do you mean using the trait approach? If so, sure, I am doing. If not, please let me know. Thanks.

mgaido91 committed:
use maropu's approach

SparkQA commented:
**[Test build #102147 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102147/testReport)** for PR 22957 at commit [`09b9981`](https://github.com/apache/spark/commit/09b99815e6cd34fd58d4fbee30b2cebf2b863d10).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7788/
Test PASSed.

SparkQA commented:
**[Test build #102147 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102147/testReport)** for PR 22957 at commit [`09b9981`](https://github.com/apache/spark/commit/09b99815e6cd34fd58d4fbee30b2cebf2b863d10).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait AliasAwareOutputPartitioning extends UnaryExecNode `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102147/
Test FAILed.

mgaido91 committed:
fix

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7791/
Test PASSed.

SparkQA commented:
**[Test build #102150 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102150/testReport)** for PR 22957 at commit [`69f9d5e`](https://github.com/apache/spark/commit/69f9d5e74d87a9d08e24fdcbe3926b066fe5a3a3).

SparkQA commented:
**[Test build #102150 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102150/testReport)** for PR 22957 at commit [`69f9d5e`](https://github.com/apache/spark/commit/69f9d5e74d87a9d08e24fdcbe3926b066fe5a3a3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102150/
Test FAILed.

mgaido91 committed:
fix ut failures

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7792/
Test PASSed.

SparkQA commented:
**[Test build #102154 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102154/testReport)** for PR 22957 at commit [`75ef545`](https://github.com/apache/spark/commit/75ef545f32403acfb0c7b9efface2b62d8f8eb0f).

SparkQA commented:
**[Test build #102154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102154/testReport)** for PR 22957 at commit [`75ef545`](https://github.com/apache/spark/commit/75ef545f32403acfb0c7b9efface2b62d8f8eb0f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102154/
Test PASSed.

HeartSaVioR commented:
Just a 2 cents, personally I understood @maropu's comment as taking up his PR, in other words, rebasing this branch to his branch to retain his commits, and adding your commit on top of his branch to fix up remaining issue. This would make giving authorship of this PR to both of @maropu and @mgaido91 easier.

maropu commented:
Thanks, @mgaido91 and @HeartSaVioR! The fix @mgaido91 did looks ok to me. But, I don't review yet and I'll do in a few days.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputPartitioning.scala:
```diff
@@ -0,0 +1,103 @@
+ * caused by the rename of an attribute among the partitioning ones, eg.
+ *
+ * spark.range(10).selectExpr("id AS key", "0").repartition($"key").write.saveAsTable("df1")
+ * spark.range(10).selectExpr("id AS key", "0").repartition($"key").write.saveAsTable("df2")
```
do you mean view here? otherwise the physical plan doesn't match.

mgaido91 commented:
@HeartSaVioR @maropu rebasing this PR with @maropu 's change and work on it is non-trivial, I can rather close this and create a new one based on @maropu 's branch if you prefer. Otherwise committers can add @maropu as author when this will be merged or @maropu can take this over and create a new PR based on this. Just let me know how you prefer to go on, thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -235,6 +241,14 @@ case class HashPartitioning(expressions: Seq[Expression], numPartitions: Int)
+
+  override private[spark] def pruneInvalidAttribute(invalidAttr: Attribute): Partitioning = {
+    if (this.references.contains(invalidAttr)) {
+      UnknownPartitioning(numPartitions)
```
`HashPartitioning('a, 'b)` with output expressions `'a as 'a1`, should produce `UnknownPartitioning` or `HashPartitioning('a1)`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -284,6 +298,19 @@ case class RangePartitioning(ordering: Seq[SortOrder], numPartitions: Int)
+
+  override private[spark] def pruneInvalidAttribute(invalidAttr: Attribute): Partitioning = {
+    if (this.references.contains(invalidAttr)) {
+      val validExprs = this.children.takeWhile(!_.references.contains(invalidAttr))
```
`this.children` -> `ordering`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:
```diff
@@ -284,6 +298,19 @@ case class RangePartitioning(ordering: Seq[SortOrder], numPartitions: Int)
+      if (validExprs.isEmpty) {
+        UnknownPartitioning(numPartitions)
+      } else {
+        RangePartitioning(validExprs, numPartitions)
```
think about `RangePartitioning('a.ASC, 'b.ASC)` with output expression `'a as 'a1`.

It cannot satisfy `ClusteredDistribution('a1)`, but can still satisfy `OrderedDistribution('a1.ASC)`. I think the expected result should be `RangePartitioning('a1.ASC, 'b.ASC)`.

mgaido91 replied:
Why doesn't it satisfy `ClusteredDistribution('a1)`? I don't agree with what you stated. If `b` is not in the output, it is useless to have it there. Moreover, when ordering we order always for the first attribute, then for the second, ... So if something is partitioned by `RangePartitioning('a1.ASC, 'b.ASC)`, it is also true that its partitioning is `RangePartitioning('a1.ASC, )`. So I think that in that case `RangePartitioning('a1.ASC)` is the right one.

cloud-fan replied:
according to `RangePartitioning.satisfies0`, and the comments in `RangePartitioning` and `ClusteredDistribution`, `RangePartitioning('a.ASC, 'b.ASC)` does not satisfy `ClusteredDistribution('a)`.

mgaido91 replied:
mmmh, I am not sure about what you mean referring to the classdoc of these 2 classes. I see nothing there about this. Anyway, I see that the implementation is done according what you state, but I do believe that it is wrong (or at least suboptimal if you prefer). If the data is partitioned by sorting it with `a.ASC, b.ASC`, it is definitely partitioned by sorting it with `a.ASC`. I think that `forall` should be an `exists`. There is also a (very minor) bug in the current implementation; try and running this test (it fails...):
```
  test("partitioning test") {
    val attr1 = AttributeReference("attr1", IntegerType)()
    val attr2 = AttributeReference("attr2", IntegerType)()
    val partitioning = RangePartitioning(Seq.empty, 10)
    val requiredDistribution = ClusteredDistribution(Seq(attr2, attr1), Some(10))
    assert(!partitioning.satisfies(requiredDistribution))
  }
```

cloud-fan replied:
please carefully read the classdoc of `RangePartitioning` and `ClusteredDistribution`, and see what `RangePartitioning` guarantees and what `ClusteredDistribution` requires.
```
partition 1: (a=1, b=2), (a=1, b=3)
partition 2: (a=1, b=4), (a=1, b=5)
```
This data set is range partitioned by a,b, but not clustered by a.

mgaido91 replied:
ah I see now, thanks. Let me update it then accordingly, thanks.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/AliasAwareOutputPartitioning.scala:
```diff
@@ -0,0 +1,103 @@
+  final override def outputPartitioning: Partitioning = {
+    child.outputPartitioning match {
+      case partitioning: Expression =>
+        val exprToEquiv = partitioning.references.map { attr =>
```
can you explain what's going on here? The code is a little hard to follow.

mgaido91 replied:
sure, let me add some comments. Thanks.

mgaido91 committed:
adress comments

SparkQA commented:
**[Test build #102193 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102193/testReport)** for PR 22957 at commit [`df3394c`](https://github.com/apache/spark/commit/df3394c010d3d58893dca887262a68c5a70e1fee).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7810/
Test PASSed.

SparkQA commented:
**[Test build #102193 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102193/testReport)** for PR 22957 at commit [`df3394c`](https://github.com/apache/spark/commit/df3394c010d3d58893dca887262a68c5a70e1fee).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102193/
Test PASSed.

mgaido91 committed:
fix rangepartitioning

SparkQA commented:
**[Test build #102201 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102201/testReport)** for PR 22957 at commit [`78d92bc`](https://github.com/apache/spark/commit/78d92bce7b56ff33f2a509ab6f7add5706686e00).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/7817/
Test PASSed.

SparkQA commented:
**[Test build #102201 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/102201/testReport)** for PR 22957 at commit [`78d92bc`](https://github.com/apache/spark/commit/78d92bce7b56ff33f2a509ab6f7add5706686e00).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/102201/
Test PASSed.

mgaido91 committed:
Merge branch 'master' into SPARK-25951

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9814/
Test PASSed.

SparkQA commented:
**[Test build #104484 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104484/testReport)** for PR 22957 at commit [`47a8f71`](https://github.com/apache/spark/commit/47a8f71b8358e9775d012dcbab6f2f03057aed94).

SparkQA commented:
**[Test build #104484 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104484/testReport)** for PR 22957 at commit [`47a8f71`](https://github.com/apache/spark/commit/47a8f71b8358e9775d012dcbab6f2f03057aed94).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104484/
Test PASSed.

mgaido91 commented:
@cloud-fan @maropu sorry for pinging you again, I think I addressed all your comments on this, may you please check it again? Thanks.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/15726/
Test PASSed.
